clang/utils/TableGen/ClangDiagnosticsEmitter.cpp

```diff
@@ -1908,7 +1908,7 @@ void clang::EmitClangDiagDocs(const RecordKeeper &Records, raw_ostream &OS) {
   for (const Record *G : DiagGroups) {
     bool IsRemarkGroup = isRemarkGroup(G, DiagsInGroup);
     auto &GroupInfo =
-        DiagsInGroup[std::string(G->getValueAsString("GroupName"))];
```

Comment:
Maybe replace with `llvm::StringMap` to get better performance and avoid these problems for good?
I would be surprised if the `StringRef` in the map is critical for performance and I am 99% certain that `StringMap` will likely win over `std::map<StringRef, ...` anyway.

Reply:
Interesting, there is a recent commit which changes the `map<std::string, ...>` to `map<llvm::StringRef, ...>`, https://github.com/llvm/llvm-project/commit/17c6ec6db1430e7e00c0e2a2ad6d26fa94fe8cf1 , which leads to this dangling reference, cc @jurahul

If we want to change to StringMap, I think it is a separate change.

Reply:
Agreed, changing to StringMap would be a different change. Also, we need to check if that ok to do in case we iterate over the map in a way that requires deterministic iteration order (and if StringMap does not guarantee that).

Can you fix the clang-format issue before approving?

Reply:
> Can you fix the clang-format issue before approving?

Done.

Reply:
If we need to have stable order, let's document it. Could you elaborate what is "different" about a change to StringMap?

It's a container that does key-value lookups with string keys. It is efficient, widely used in LLVM and fully avoids this class of errors. Are there any downsides of using it (assuming we don't need to stable iteration order that std::map provides?)

Reply:
Yeah, we need to document it. I can see from the code that we iterate over the map (in emitDiagTable)

```
  for (auto const &[Name, GroupInfo] : DiagsInGroup) {
```

with std::map, this iteration happens in an order sorted by the keys in the map. And that is needed so that the tablegen generated code is deterministic. I don't think StringMap similar iteration order guarantees. See https://llvm.org/docs/ProgrammersManual.html#llvm-adt-stringmap-h

```
StringMap iteration order, however, is not guaranteed to be deterministic, so any uses which require that should instead use a std::map.
```